File: classifier/naivebayesclassifier.py
from collections import defaultdict
from math import log, exp, log10


class NaiveBayesClassifier:
    def __init__(self):
        self.frequency_table = defaultdict(lambda: 0)
        self.messages_in_class = defaultdict(lambda: 0)
        self.messages_total = 0
        self.words_in_class = defaultdict(lambda: 0)
        self.words = defaultdict(lambda: 0)

    def train(self, samples):
        for msg in range(len(samples)):
            features = samples[msg][0]
            label = samples[msg][1]
            self.messages_in_class[label] += 1
            self.messages_total += 1
            for feature in features:
                self.frequency_table[label, feature] += 1
                self.words_in_class[label] += 1
                self.words[feature] += 1

    def classify(self, features, delim=0.5):
        keys = {x: log(self.messages_in_class[x] / self.messages_total)
                        + sum([log((self.frequency_table[x, feat] + 1)/(len(self.words) + self.words_in_class[x]))
                        for feat in features]) for x in self.messages_in_class.keys()}
        degree = keys[1] - keys[0]
        degree = -log(-degree) if degree < 0 else log(degree) if degree > 0 else 0
        pos = 0 if degree > 700 else 1 if degree < -700 else 1 / (1 + exp(degree))
        return 0 if pos > delim else 1

File: src/classifier/NaiveBayesClassifier.java
package classifier;

import javafx.util.Pair;
import lib.extra.Sample;
import lib.python.collections.DefaultDict;

import java.util.List;

public class NaiveBayesClassifier implements Classifier {
    private DefaultDict<Pair<Integer, Integer>> frequencyTable;
    private DefaultDict<Integer> messagesInClass;
    private int messagesTotal;
    private DefaultDict<Integer> wordsInClass;
    private DefaultDict<Integer> words;

    public NaiveBayesClassifier(){
        refresh();
    }

    @Override
    public void refresh(){
        this.frequencyTable = new DefaultDict<>(0);
        this.messagesInClass = new DefaultDict<>(0);
        this.messagesTotal = 0;
        this.wordsInClass = new DefaultDict<>(0);
        this.words = new DefaultDict<>(0);
    }

    @Override
    public void train(List<Sample> samples){
        for (Sample sample : samples){
            this.messagesInClass.increment(sample.label);
            this.messagesTotal++;
            for (Integer feature : sample.features){
                this.frequencyTable.increment(new Pair<>(sample.label, feature));
                this.wordsInClass.increment(sample.label);
                this.words.increment(feature);
            }
        }
    }

    private double formula(Integer label, List<Integer> features){
        double sum = 0;
        for (Integer feature : features){
            sum += Math.log((this.frequencyTable.get(new Pair<>(label, feature)) + 1)) /
                    (this.words.size() + this.wordsInClass.get(label));
        }
        return Math.log(this.messagesInClass.get(label) / this.messagesTotal) + sum;
    }

    @Override
    public int classify(List<Integer> features){
        int maxKey = 0;
        double maxVal = Double.MIN_VALUE;
        for (Integer label : this.messagesInClass.keys()){
            double value = formula(label, features);
            if (maxKey == -1 || value > maxVal) {
                maxKey = label;
                maxVal = value;
            }
        }
        return maxKey;
    }
}
